File: xen/brain/memory.py
"""Xen's memories: everyday experience plus vivid memories of what hurt and what delighted.

Experience is remembered as short stretches (n steps), so the harm or reward
at the end of a stretch is linked to the choices at its start: the cue before
the shock.  Painful stretches go into *trauma* memory and very rewarding ones
into *joy* memory; both are replayed far more often than ordinary moments,
which is how a single burn teaches Xen to fear lava (fear conditioning), and a
single diamond makes digging deep feel worth it.
"""
from collections import defaultdict, deque

import numpy as np


class Replay:
    """Fixed-size ring buffer of transitions (observations stored as int8).

    Each entry holds both views of a moment: what happened one step later
    (reward1, harm, next1, done1) and the discounted stretch of n steps
    (reward, next_obs, done, steps).
    """

    FIELDS = ("action", "reward", "done", "steps", "reward1", "harm", "done1")

    def __init__(self, capacity, obs_dim):
        self.capacity = int(capacity)
        self.obs = np.zeros((self.capacity, obs_dim), np.int8)
        self.next_obs = np.zeros((self.capacity, obs_dim), np.int8)
        self.next1 = np.zeros((self.capacity, obs_dim), np.int8)
        for name in self.FIELDS:
            setattr(self, name, np.zeros(self.capacity, np.int64 if name == "action" else np.float32))
        self.size = 0
        self._next = 0

    @staticmethod
    def _pack(obs):
        return np.round(np.clip(obs, -1, 1) * 127).astype(np.int8)

    def add(self, obs, action, reward, next_obs, done, steps, reward1, harm, next1, done1):
        i = self._next
        self.obs[i] = self._pack(obs)
        self.next_obs[i] = self._pack(next_obs)
        self.next1[i] = self._pack(next1)
        for name, value in zip(self.FIELDS, (action, reward, float(done), steps, reward1, harm, float(done1))):
            getattr(self, name)[i] = value
        self._next = (i + 1) % self.capacity
        self.size = min(self.size + 1, self.capacity)

    def take(self, idx):
        batch = {name: getattr(self, name)[idx] for name in self.FIELDS}
        batch["obs"] = self.obs[idx].astype(np.float32) / 127.0
        batch["next_obs"] = self.next_obs[idx].astype(np.float32) / 127.0
        batch["next1"] = self.next1[idx].astype(np.float32) / 127.0
        return batch

    def __len__(self):
        return self.size


class Memory:
    """Ordinary experience + trauma + joy, sampled together."""

    def __init__(self, obs_dim, capacity=50_000, trauma_capacity=5_000, joy_capacity=5_000,
                 n_step=3, gamma=0.97, fear_gamma=0.9, trauma_fraction=0.2, joy_fraction=0.1,
                 joy_threshold=0.5, seed=0):
        self.experience = Replay(capacity, obs_dim)
        self.trauma = Replay(trauma_capacity, obs_dim)
        self.joy = Replay(joy_capacity, obs_dim)
        self.n_step = n_step
        self.gamma, self.fear_gamma = gamma, fear_gamma
        self.trauma_fraction, self.joy_fraction = trauma_fraction, joy_fraction
        self.joy_threshold = joy_threshold
        # One short-term memory per body, so a swarm of Xens sharing a brain
        # don't mix up each other's moments.
        self._recent = defaultdict(lambda: deque(maxlen=n_step))
        self._rng = np.random.default_rng(seed)

    def remember(self, obs, action, reward, harm, next_obs, done, stream=0, end=False):
        """done: the life ended (death); end: the stretch stops here anyway (time's up)."""
        recent = self._recent[stream]
        recent.append((obs, action, reward, harm, next_obs, done))
        if len(recent) == self.n_step:
            self._commit(recent, next_obs, done)
        if done or end:
            while recent:
                self._commit(recent, next_obs, done)
            del self._recent[stream]

    def _commit(self, recent, next_obs, done):
        obs, action, reward1, harm1, next1, done1 = recent[0]
        reward = sum(self.gamma ** i * t[2] for i, t in enumerate(recent))
        harm_ahead = sum(t[3] for t in recent)
        entry = (obs, action, reward, next_obs, done, len(recent), reward1, harm1, next1, done1)
        self.experience.add(*entry)
        if harm_ahead > 0:
            self.trauma.add(*entry)          # the cue before the shock, and the shock
        if reward >= self.joy_threshold:
            self.joy.add(*entry)
        recent.popleft()

    def sample(self, n):
        n_trauma = min(int(round(n * self.trauma_fraction)), len(self.trauma))
        n_joy = min(int(round(n * self.joy_fraction)), len(self.joy))
        parts = [self.experience.take(self._rng.integers(0, len(self.experience), n - n_trauma - n_joy))]
        if n_trauma:
            parts.append(self.trauma.take(self._rng.integers(0, len(self.trauma), n_trauma)))
        if n_joy:
            parts.append(self.joy.take(self._rng.integers(0, len(self.joy), n_joy)))
        return {k: np.concatenate([p[k] for p in parts]) for k in parts[0]}

    def __len__(self):
        return len(self.experience)

File: mod/common/java/xen/mod/core/Memory.java
package xen.mod.core;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

/** Experience + trauma + joy memories of n-step stretches. Port of xen/brain/memory.py. */
public final class Memory {
	/** One remembered moment: both the next step and the stretch of n steps after it. */
	public static final class Entry {
		byte[] obs, next, next1;
		int action;
		float reward, done, steps, reward1, harm, done1;
	}

	static final class Ring {
		final Entry[] items;
		int size, next;

		Ring(int capacity) {
			items = new Entry[capacity];
		}

		void add(Entry e) {
			items[next] = e;
			next = (next + 1) % items.length;
			size = Math.min(size + 1, items.length);
		}
	}

	final Ring experience, trauma, joy;
	final int nStep;
	final float gamma, traumaFraction, joyFraction, joyThreshold = 0.5f;
	private final Map<Object, ArrayDeque<Object[]>> recent = new HashMap<>();
	private final Random rng = new Random(3);

	public Memory(int capacity, int traumaCapacity, int nStep, float gamma, float traumaFraction, float joyFraction) {
		experience = new Ring(capacity);
		trauma = new Ring(traumaCapacity);
		joy = new Ring(traumaCapacity);
		this.nStep = nStep;
		this.gamma = gamma;
		this.traumaFraction = traumaFraction;
		this.joyFraction = joyFraction;
	}

	static byte[] pack(float[] o) {
		byte[] b = new byte[o.length];
		for (int i = 0; i < o.length; i++) b[i] = (byte) Math.round(Math.max(-1, Math.min(1, o[i])) * 127);
		return b;
	}

	static float[] unpack(byte[] b) {
		float[] o = new float[b.length];
		for (int i = 0; i < b.length; i++) o[i] = b[i] / 127f;
		return o;
	}

	public int size() {
		return experience.size;
	}

	public int traumaSize() {
		return trauma.size;
	}

	public int joySize() {
		return joy.size;
	}

	/** Vivid memories go with the brain when it's saved, so it never forgets what hurt it (or delighted it). */
	static final int KEEP = 1000;

	/** The most recent n entries of a ring, oldest first. */
	static List<Entry> recent(Ring r, int n) {
		List<Entry> out = new ArrayList<>();
		n = Math.min(n, r.size);
		for (int k = n; k > 0; k--) out.add(r.items[Math.floorMod(r.next - k, r.items.length)]);
		return out;
	}

	Ring ring(String name) {
		return switch (name) {
			case "trauma" -> trauma;
			case "joy" -> joy;
			default -> throw new IllegalArgumentException("unknown memory " + name);
		};
	}

	/** Write entries (same layout as xen/brain/memory.py: int8 observations, then int32 actions, then float32s). */
	static void write(List<Entry> entries, int obsDim, java.io.OutputStream out) throws java.io.IOException {
		int n = entries.size();
		for (int part = 0; part < 3; part++) {
			byte[] b = new byte[n * obsDim];
			for (int i = 0; i < n; i++) {
				Entry e = entries.get(i);
				System.arraycopy(part == 0 ? e.obs : part == 1 ? e.next : e.next1, 0, b, i * obsDim, obsDim);
			}
			out.write(b);
		}
		java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(n * 4 * 7).order(java.nio.ByteOrder.LITTLE_ENDIAN);
		for (Entry e : entries) bb.putInt(e.action);
		for (Entry e : entries) bb.putFloat(e.reward);
		for (Entry e : entries) bb.putFloat(e.done);
		for (Entry e : entries) bb.putFloat(e.steps);
		for (Entry e : entries) bb.putFloat(e.reward1);
		for (Entry e : entries) bb.putFloat(e.harm);
		for (Entry e : entries) bb.putFloat(e.done1);
		out.write(bb.array());
	}

	static void read(Ring r, int n, int obsDim, java.io.DataInputStream in) throws java.io.IOException {
		Entry[] es = new Entry[n];
		for (int i = 0; i < n; i++) es[i] = new Entry();
		for (int part = 0; part < 3; part++) {
			byte[] b = new byte[n * obsDim];
			in.readFully(b);
			for (int i = 0; i < n; i++) {
				byte[] o = java.util.Arrays.copyOfRange(b, i * obsDim, (i + 1) * obsDim);
				if (part == 0) es[i].obs = o;
				else if (part == 1) es[i].next = o;
				else es[i].next1 = o;
			}
		}
		byte[] raw = new byte[n * 4 * 7];
		in.readFully(raw);
		java.nio.ByteBuffer bb = java.nio.ByteBuffer.wrap(raw).order(java.nio.ByteOrder.LITTLE_ENDIAN);
		for (Entry e : es) e.action = bb.getInt();
		for (Entry e : es) e.reward = bb.getFloat();
		for (Entry e : es) e.done = bb.getFloat();
		for (Entry e : es) e.steps = bb.getFloat();
		for (Entry e : es) e.reward1 = bb.getFloat();
		for (Entry e : es) e.harm = bb.getFloat();
		for (Entry e : es) e.done1 = bb.getFloat();
		for (Entry e : es) r.add(e);
	}

	public void remember(float[] obs, int action, float reward, float harm, float[] next, boolean done, Object stream, boolean end) {
		ArrayDeque<Object[]> q = recent.computeIfAbsent(stream, k -> new ArrayDeque<>());
		q.addLast(new Object[] {obs, action, reward, harm, next, done});
		if (q.size() == nStep) commit(q, next, done);
		if (done || end) {
			while (!q.isEmpty()) commit(q, next, done);
			recent.remove(stream);
		}
	}

	private void commit(ArrayDeque<Object[]> q, float[] next, boolean done) {
		Object[] first = q.peekFirst();
		float reward = 0, harmAhead = 0;
		int i = 0;
		for (Object[] t : q) {
			reward += (float) Math.pow(gamma, i++) * (float) t[2];
			harmAhead += (float) t[3];
		}
		Entry e = new Entry();
		e.obs = pack((float[]) first[0]);
		e.action = (int) first[1];
		e.reward1 = (float) first[2];
		e.harm = (float) first[3];
		e.next1 = pack((float[]) first[4]);
		e.done1 = (boolean) first[5] ? 1 : 0;
		e.reward = reward;
		e.next = pack(next);
		e.done = done ? 1 : 0;
		e.steps = q.size();
		experience.add(e);
		if (harmAhead > 0) trauma.add(e);
		if (reward >= joyThreshold) joy.add(e);
		q.pollFirst();
	}

	public List<Entry> sample(int n) {
		int nT = Math.min(Math.round(n * traumaFraction), trauma.size);
		int nJ = Math.min(Math.round(n * joyFraction), joy.size);
		List<Entry> out = new ArrayList<>(n);
		for (int i = 0; i < n - nT - nJ; i++) out.add(experience.items[rng.nextInt(experience.size)]);
		for (int i = 0; i < nT; i++) out.add(trauma.items[rng.nextInt(trauma.size)]);
		for (int i = 0; i < nJ; i++) out.add(joy.items[rng.nextInt(joy.size)]);
		return out;
	}
}
